cmd/main.go

```diff
@@ -0,0 +1,128 @@
+)
+
+func Run() {
+
+	promlogConfig := &promlog.Config{}
+
+	flag.AddFlags(kingpin.CommandLine, promlogConfig)
+	kingpin.Version(version.Print("json_exporter"))
+	kingpin.HelpFlag.Short('h')
+	kingpin.Parse()
+	logger := promlog.New(promlogConfig)
+
+	level.Info(logger).Log("msg", "Starting json_exporter", "version", version.Info()) //nolint:errcheck
+	level.Info(logger).Log("msg", "Build context", "build", version.BuildContext())    //nolint:errcheck
+
+	level.Info(logger).Log("msg", "Loading config file", "file", *configFile) //nolint:errcheck
+	config, err := config.LoadConfig(*configFile)
+	if err != nil {
+		level.Error(logger).Log("msg", "Error loading config", "err", err) //nolint:errcheck
+		os.Exit(1)
+	}
+	configJson, err := json.Marshal(config)
+	if err != nil {
+		level.Error(logger).Log("msg", "Failed to marshal config to JOSN", "err", err) //nolint:errcheck
+	}
+	level.Info(logger).Log("msg", "Loaded config file", "config", string(configJson)) //nolint:errcheck
+
+	if *configCheck {
+		os.Exit(0)
+	}
+
+	http.Handle("/metrics", promhttp.Handler())
+	http.HandleFunc("/probe", func(w http.ResponseWriter, req *http.Request) {
+		probeHandler(w, req, logger, config)
+	})
+	if err := http.ListenAndServe(*listenAddress, nil); err != nil {
+		level.Error(logger).Log("msg", "failed to start the server", "err", err) //nolint:errcheck
+	}
+}
+
+func probeHandler(w http.ResponseWriter, r *http.Request, logger log.Logger, config config.Config) {
+
+	ctx, cancel := context.WithTimeout(r.Context(), time.Duration(config.Global.TimeoutSeconds*float64(time.Second)))
+	defer cancel()
+	r = r.WithContext(ctx)
+
+	registry := prometheus.NewPedanticRegistry()
+
+	metrics, err := internal.CreateMetricsList(registry, config)
+	if err != nil {
+		level.Error(logger).Log("msg", "Failed to create metrics list from config", "err", err) //nolint:errcheck
+	}
+
+	probeSuccessGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_success",
+		Help: "Displays whether or not the probe was a success",
+	})
+	probeDurationGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_duration_seconds",
+		Help: "Returns how long the probe took to complete in seconds",
+	})
+
+	target := r.URL.Query().Get("target")
+	if target == "" {
+		http.Error(w, "Target parameter is missing", http.StatusBadRequest)
+		return
+	}
+
+	start := time.Now()
+	registry.MustRegister(probeSuccessGauge)
+	registry.MustRegister(probeDurationGauge)
+
+	data, err := internal.FetchJson(ctx, logger, target, config.Headers)
+	if err != nil {
+		level.Error(logger).Log("msg", "Failed to fetch JSON response", "err", err) //nolint:errcheck
+		duration := time.Since(start).Seconds()
+		level.Error(logger).Log("msg", "Probe failed", "duration_seconds", duration) //nolint:errcheck
+	} else {
+		internal.Scrape(logger, metrics, data)
+
+		duration := time.Since(start).Seconds()
+		probeDurationGauge.Set(duration)
+		probeSuccessGauge.Set(1)
```

Comment:
This is not going to work. The first success will set to 1, and then never un-set. In order to do what you want, you need to use a const metric. See the way we refactored the pgbouncer_exporter.

https://github.com/prometheus-community/pgbouncer_exporter/pull/33

Reply:
Well, I was facing this problem earlier, but not only for `probe_success`, but for all the metrics in the registry. Adding the `context` to the `probeHandler` function fixed it.

Reply:
Yes, but this is a very non-standard way to do this. The standard way for exporters like to expose data is to use the Const metric pattern.

Reply:
I'd expect this scrape to fail completely if it can't fetch the json

Reply:
> Yes, but this is a very non-standard way to do this. The standard way for exporters like to expose data is to use the Const metric pattern.

Yes, that sounds fair. Initially when I started working on this exporter I didn't quite understand the Const Metric concept so didn't use it. I think I understand that a little better now, so will switch to Const Metric pattern.

> I'd expect this scrape to fail completely if it can't fetch the json

Do you mean that instead of the current `probe_success=0` response, should it just forward whatever error it receives from the JSON endpoint, as it is to prometheus?

Reply:
We would return a 502 bad gateway or something like this, yes.

Reply:
Maybe 503?

Reply:
yes :+1:

Reply:
Fixed both of the issues as recommended.